File: cpp-implementation/src/sphere.cpp
module sphere;
import std;
namespace cg
{

std::optional<hitevent> get_ray_sphere_hit(const sphere& s, ray r, interval i)
{
  const auto dir_to_sphere = s.pos - r.pos;
  auto a = glm::length2(r.dir);
  auto h = glm::dot(r.dir, dir_to_sphere);
  auto c = glm::length2(dir_to_sphere) - s.radius * s.radius;

  auto discriminant = h * h - a * c;

  if (discriminant < 0.f)
  {
    return std::nullopt;
  }

  const float sqrt_discriminant = std::sqrt(discriminant);
  float t = (h - sqrt_discriminant) / a;
  if (t < 0.f || !i.contains(t))
  {
    t = (h + sqrt_discriminant) / a;
    if (t < 0.f || !i.contains(t))
      return std::nullopt;
  }
  auto normal = glm::normalize((r.at(t) - s.pos) / s.radius);
  if (glm::dot(normal, r.dir) > 0.f)
  {
    normal = -normal;
  }
  return hitevent{t, r.at(t), normal, s.material_id};
}

} // namespace cg

File: class-oop-implementation/src/objects/Sphere.java
package objects;
import java.util.Optional;
import math.*;

public class Sphere implements Object3D {
  private final Vector3D center;
  private final float radius;
  private final int materialId;

  public Sphere(Vector3D center, float radius, int materialId) {
    this.center = center;
    this.radius = radius;
    this.materialId = materialId;
  }

  public Vector3D getCenter() {
    return center;
  }

  public float getRadius() {
    return radius;
  }

  @Override
  public Optional<Intersection> isHit(Ray ray, Interval tRange) {
    Vector3D dirToSphere = center.sub(ray.getPos());
    float a = ray.getDir().dot(ray.getDir());
    float h = ray.getDir().dot(dirToSphere);
    float c = dirToSphere.dot(dirToSphere) - radius * radius;

    float discriminant = h * h - a * c;
    if (discriminant < 0) {
      return Optional.empty();
    }

    float sqrtDiscriminant = (float) Math.sqrt(discriminant);
    float t = (h - sqrtDiscriminant) / a;
    if (t < 0 || !tRange.contains(t)) {
      t = (h + sqrtDiscriminant) / a;
      if (t < 0 || !tRange.contains(t)) {
        return Optional.empty();
      }
    }

    Vector3D hitPoint = ray.at(t);
    Vector3D normal = hitPoint.sub(center).mul(1f / radius).normalize();
    if (normal.dot(ray.getDir()) > 0f) {
      normal = normal.mul(-1f);
    }
    return Optional.of(new Intersection(hitPoint, normal, t, materialId));
  }

}
